File: leetcode/0211-design-add-and-search-words-data-structure/solution.cpp
#include <string>
#include <unordered_map>
using namespace std;

struct TrieNode {
    unordered_map<char, TrieNode*> children;
    bool isEnd = false;
};

class WordDictionary {
public:
    WordDictionary() {
        root = new TrieNode();
    }

    void addWord(string word) {
        TrieNode* node = root;
        for (char c : word) {
            if (!node->children.count(c)) node->children[c] = new TrieNode();
            node = node->children[c];
        }
        node->isEnd = true;
    }

    bool search(string word) {
        return dfs(root, word, 0);
    }

private:
    TrieNode* root;

    bool dfs(TrieNode* node, const string& word, int i) {
        if (i == (int)word.size()) return node->isEnd;

        char c = word[i];
        if (c == '.') {
            for (auto& [ch, child] : node->children) {
                if (dfs(child, word, i + 1)) return true;
            }
            return false;
        }

        auto it = node->children.find(c);
        return it != node->children.end() && dfs(it->second, word, i + 1);
    }
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary* obj = new WordDictionary();
 * obj->addWord(word);
 * bool param_2 = obj->search(word);
 */

File: leetcode/0211-design-add-and-search-words-data-structure/Solution.java
import java.util.HashMap;
import java.util.Map;

class WordDictionary {
    static class TrieNode {
        Map<Character, TrieNode> children = new HashMap<>();
        boolean isEnd = false;
    }

    private final TrieNode root;

    public WordDictionary() {
        root = new TrieNode();
    }

    public void addWord(String word) {
        TrieNode node = root;
        for (char c : word.toCharArray()) {
            node = node.children.computeIfAbsent(c, k -> new TrieNode());
        }
        node.isEnd = true;
    }

    public boolean search(String word) {
        return dfs(root, word, 0);
    }

    private boolean dfs(TrieNode node, String word, int i) {
        if (i == word.length()) {
            return node.isEnd;
        }

        char c = word.charAt(i);
        if (c == '.') {
            for (TrieNode child : node.children.values()) {
                if (dfs(child, word, i + 1)) {
                    return true;
                }
            }
            return false;
        }

        TrieNode child = node.children.get(c);
        return child != null && dfs(child, word, i + 1);
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */
